v3/pforzheim.py

```diff
@@ -0,0 +1,113 @@
+
+
+class PforzheimConverter(JsonConverter):
+    pforzheim_row_validator = DataclassValidator(PforzheimRowInput)
+
+    source_info = SourceInfo(
+        id='pforzheim',
+        name='Stadt Pforzheim',
+        public_url='https://www.pforzheim.de',
+    )
+
+    type_mapping: dict[str, ParkingSiteTypeInput] = {
+        'onStreet': ParkingSiteTypeInput.ON_STREET,
+        'carPark': ParkingSiteTypeInput.CAR_PARK,
+        'undergroundCarPark': ParkingSiteTypeInput.UNDERGROUND,
+    }
+
+    def handle_json(self, data: dict | list) -> ImportSourceResult:
+        import_source_result = self.generate_import_source_result(
+            static_parking_site_inputs=[],
+            static_parking_site_errors=[],
+        )
+
+        # We start at row 2, as the first one is our header
+        for item in data:
+            input_dict = {
+                'uid': item.get('Id'),
+                'name': item.get('name'),
+                'lat': str(item.get('lat')),
+                'lon': str(item.get('lon')),
+                'operator_name': item.get('operatorID'),
+                'address': item.get('address'),
+                'description': item.get('description'),
+                'type': item.get('type'),
+                'capacity_woman': item.get('quantitySpacesReservedForWomen'),
+                'capacity_disabled': item.get('quantitySpacesReservedForMobilityImpededPerson'),
+                'is_supervised': True if 'ja' in item.get('securityInformation').lower() else False,
+                'fee_description': item.get('feeInformation'),
+                'capacity': item.get('capacity'),
+                'opening_hours': item.get('openingHours'),
+            }
+
+            try:
+                input_data: PforzheimRowInput = self.pforzheim_row_validator.validate(input_dict)
+            except ValidationError as e:
+                import_source_result.static_parking_site_errors.append(
+                    ImportParkingSiteException(
+                        uid=input_dict.get('id'),
+                        message=f'validation error for {input_dict}: {e.to_dict()}',
+                    ),
+                )
+                continue
+
+            parking_site_input = StaticParkingSiteInput(
+                uid=name_to_id(self.source_info.id, input_data.name) if input_data.uid == '' else input_data.uid,
+                name=input_data.name,
+                type=self.type_mapping.get(input_data.type),
+                lat=input_data.lat,
+                lon=input_data.lon,
+                address=input_data.address.replace('\n', ' '),
+                description=input_data.description,
+                capacity=input_data.capacity,
+                capacity_woman=input_data.capacity_woman,
+                capacity_disabled=input_data.capacity_disabled,
+                fee_description=input_data.fee_description,
+                opening_hours='24/7' if input_data.opening_hours == 'durchgehend geöffnet' else None,
```

Comment:
There is a field called `hasOpeningHours24h` in our source data which is a mapped boolean and therefore better than a string content comparsion.

Reply:
At first, when I add ```hasOpeningHours24h```  to the converter, I get the following error report:
```
Traceback (most recent call last):
  File "ParkAPI2-sources/test-push-converter.py", line 125, in <module>
    main()
  File "ParkAPI2-sources/test-push-converter.py", line 75, in main
    result: ImportSourceResult = converter.handle_json(data)
  File "ParkAPI2-sources/v3/pforzheim.py", line 80, in handle_json
    parking_site_input = StaticParkingSiteInput(
TypeError: StaticParkingSiteInput.__init__() got an unexpected keyword argument 'hasOpeningHours24h'
```

After adding ```hasOpeningHours24h: Optional[bool] = Noneable(BooleanValidator()), Default(None)``` to the ```base_validator.py```, it resolved.

Is it the right way?

I also noticed ```hasOpeningHours24h``` is not in the [park-api documentation](https://api.mobidata-bw.de/park-api/documentation/public.html#/paths/v3-parking-sites-parking_site_id/get)

Reply:
You added a validator to the output, not to the source validation. Your data source validation is this opne: https://github.com/mobidata-bw/ParkAPI2-sources/blob/pforzheim-push-converter/v3/pforzheim.py#L27 .

`hasOpeningHours24h` is already covered by `opening_hours` at park-apis model, and it would also not fit in our schema as its camelCase, not snake_case.

My point was that it's always better to use a well-defined field for validation at input instead of making a string-comparsion. `hasOpeningHours24h` is your source data is excaly that and replaces `if input_data.opening_hours == 'durchgehend geöffnet'`.

Reply:
Now I found the ```hasOpenHours24h``` in the source data. It was an overseeing on my side, I would have preferred using it as well since it is boolean-related and well-defined. Thanks!